File: android-pgs-plugin/src/SignInController.java
package org.godotengine.godot;

import android.app.Activity;
import android.content.Intent;
import android.util.Log;
import android.util.Pair;
import androidx.annotation.NonNull;
import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInResult;
import com.google.android.gms.games.Games;
import com.google.android.gms.tasks.OnCompleteListener;
import com.google.android.gms.tasks.Task;

public class SignInController {

    public final static int RC_SIGN_IN = 77;

    private Activity activity;
    private GodotCallbacksUtils godotCallbacksUtils;
    private ConnectionController connectionController;
    private boolean showPlayPopups = true;

    public SignInController(Activity activity, GodotCallbacksUtils godotCallbacksUtils, ConnectionController connectionController) {
        this.activity = activity;
        this.godotCallbacksUtils = godotCallbacksUtils;
        this.connectionController = connectionController;
    }

    public void setShowPopups(boolean enablePopUps) {
        this.showPlayPopups = enablePopUps;
    }

    public void signIn(final GoogleSignInClient googleSignInClient) {
        Pair<Boolean, String> connection = connectionController.isConnected();
        if (connection.first) {
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SIGNIN_SUCCESSFUL, new Object[]{connection.second});
            enablePopUps();
        } else {
            googleSignInClient
                    .silentSignIn()
                    .addOnCompleteListener(activity, new OnCompleteListener<GoogleSignInAccount>() {
                        @Override
                        public void onComplete(@NonNull Task<GoogleSignInAccount> task) {
                            if (task.isSuccessful()) {
                                GoogleSignInAccount googleSignInAccount = task.getResult();
                                String accId = "";
					            if (googleSignInAccount != null && googleSignInAccount.getId() != null) {
					            	accId = googleSignInAccount.getId();
					            }

                                godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SIGNIN_SUCCESSFUL, new Object[]{accId});
                                enablePopUps();
                            } else {
                                Intent intent = googleSignInClient.getSignInIntent();
                                activity.startActivityForResult(intent, RC_SIGN_IN);
                            }
                        }
                    });
        }
    }

    public void onSignInActivityResult(GoogleSignInResult googleSignInResult) {
        if (googleSignInResult != null && googleSignInResult.isSuccess()) {
            GoogleSignInAccount googleSignInAccount = googleSignInResult.getSignInAccount();
            String accId = "";
            if (googleSignInAccount != null && googleSignInAccount.getId() != null) {
            	accId = googleSignInAccount.getId();
            }
            enablePopUps();
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SIGNIN_SUCCESSFUL, new Object[]{accId});
        } else {
        	int statusCode = Integer.MIN_VALUE;
        	if (googleSignInResult != null && googleSignInResult.getStatus() != null){
        		statusCode = googleSignInResult.getStatus().getStatusCode();
        	}
            godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SIGNIN_FAILED, new Object[]{statusCode});
        }
    }
    
    public void signOut(GoogleSignInClient googleSignInClient) {
        googleSignInClient.signOut().addOnCompleteListener(activity, new OnCompleteListener<Void>() {
            @Override
            public void onComplete(@NonNull Task<Void> task) {
                if (task.isSuccessful()) {
                    godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SIGN_OUT_SUCCESS, new Object[]{});
                } else {
                    godotCallbacksUtils.invokeGodotCallback(GodotCallbacksUtils.SIGN_OUT_FAILED, new Object[]{});
                }
            }
        });
    }

    private void enablePopUps() {
        if (showPlayPopups) {
            GoogleSignInAccount lastSignedInAccount = GoogleSignIn.getLastSignedInAccount(activity);
            if (lastSignedInAccount != null) {
                Games.getGamesClient(activity, lastSignedInAccount).setViewForPopups(activity.findViewById(android.R.id.content));
            }
        }
    }
}

File: app/src/main/java/io/cgisca/godot/gpgs/signin/SignInController.kt
package io.cgisca.godot.gpgs.signin

import android.app.Activity
import android.util.Log
import android.util.Pair
import com.google.android.gms.auth.api.signin.GoogleSignIn
import com.google.android.gms.auth.api.signin.GoogleSignInClient
import com.google.android.gms.auth.api.signin.GoogleSignInResult
import com.google.android.gms.games.Games
import io.cgisca.godot.gpgs.ConnectionController

class SignInController(
    private var activity: Activity,
    private var signInListener: SignInListener,
    private var connectionController: ConnectionController
) {

    companion object {
        const val RC_SIGN_IN = 77
    }

    private var showPlayPopups = true

    fun setShowPopups(enablePopUps: Boolean) {
        showPlayPopups = enablePopUps
    }

    fun signIn(googleSignInClient: GoogleSignInClient) {
        val userProfile = UserProfile(null, null, null, null)
        val connection: Pair<Boolean, UserProfile> = connectionController.isConnected()
        if (connection.first) {
            Log.i("godot","Using cached signin data")
            signInListener.onSignedInSuccessfully(connection.second)
            enablePopUps()
        } else {
            Log.i("godot","Using new signin data")
            googleSignInClient
                .silentSignIn()
                .addOnCompleteListener(activity) { task ->
                    if (task.isSuccessful) {
                        val googleSignInAccount = task.result
                        if (googleSignInAccount != null) {
                            userProfile.let {
                                it.displayName = googleSignInAccount.displayName
                                it.email = googleSignInAccount.email
                                it.token = googleSignInAccount.idToken
                                it.id = googleSignInAccount.id
                            }
                        }

                        signInListener.onSignedInSuccessfully(userProfile)
                        enablePopUps()
                    } else {
                        val intent = googleSignInClient.signInIntent
                        activity.startActivityForResult(intent, RC_SIGN_IN)
                    }
                }
        }
    }

    fun onSignInActivityResult(googleSignInResult: GoogleSignInResult?) {
        val userProfile = UserProfile(null, null, null, null)
        if (googleSignInResult != null && googleSignInResult.isSuccess) {
            val googleSignInAccount = googleSignInResult.signInAccount
            if (googleSignInAccount != null) {
                userProfile.let {
                    it.displayName = googleSignInAccount.displayName
                    it.email = googleSignInAccount.email
                    it.token = googleSignInAccount.idToken
                    it.id = googleSignInAccount.id
                }
            }
            enablePopUps()
            signInListener.onSignedInSuccessfully(userProfile)
        } else {
            var statusCode = Int.MIN_VALUE
            googleSignInResult?.status?.let {
                statusCode = it.statusCode
            }
            signInListener.onSignInFailed(statusCode)
        }
    }

    fun signOut(googleSignInClient: GoogleSignInClient) {
        googleSignInClient.signOut().addOnCompleteListener(activity) { task ->
            if (task.isSuccessful) {
                signInListener.onSignOutSuccess()
            } else {
                signInListener.onSignOutFailed()
            }
        }
    }

    fun isSignedIn(): Boolean {
        val googleSignInAccount = GoogleSignIn.getLastSignedInAccount(activity)
        return connectionController.isConnected().first && googleSignInAccount != null
    }

    private fun enablePopUps() {
        if (showPlayPopups) {
            val lastSignedInAccount = GoogleSignIn.getLastSignedInAccount(activity)
            if (lastSignedInAccount != null) {
                Games.getGamesClient(activity, lastSignedInAccount)
                    .setViewForPopups(activity.findViewById(android.R.id.content))
            }
        }
    }
}